### ISSUE TITLE
Test UI: Key Bindings do not work with contributed `testing/item/context` commands

### ISSUE THREAD

**Issue by smcenlly** [microsoft/vscode#130880]
Version: 1.60.0-insider (Universal)
Commit: 7c25c174726ab969db3f99058cdf24bcfae1cfbe
Date: 2021-08-11T05:13:17.497Z
Electron: 13.1.8
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Darwin arm64 20.5.0

Steps to Reproduce:

1. Add a key binding for a contributed `testing/item/context` command, for example:
```javascript
...
"keybindings": [
      {
        "command": "extension.command",
        "weight": 90,
        "when": "focusedView == workbench.view.testing",
        "key": "alt+F10"
      }
  ]
...
```

2. In the tree view, navigate to and select a test
3. Run the command using `alt+F10`

At this point, the command is executed, which is correct, but no parameters are provided to the command (i.e. the `TestItem` is not provided. The keyboard shortcut does work (and a test item is provided) when the context menu has first been opened.

---

Note: I believe that the same issue also affects the `alt+enter` keyboard navigation within the Test Explorer as the shortcut only appears to work after the context menu has been opened.

**Comment by TAUnionOtto** [comment]
Same issue with me.

My steps here:

1. Add view container, tree view, command,  keybinding and menu

```
{
  "contributes": {
    "commands": [
      {
        "command": "xxx.copyItemId",
        "title": "Copy ID",
        "category": "XXX Explorer"
      }
    ],
    "menus": {
      "view/item/context": [
        {
          "command": "xxx.copyItemId"
        }
      ]
    },
    "viewsContainers": {
      "activitybar": [
        {
          "id": "xxx--explorer",
          "title": "XXX Explorer",
          "icon": "assets/bank.svg"
        }
      ]
    },
    "views": {
      "xxx-explorer": [
        {
          "id": "xxxExplorer",
          "name": "XXX Explorer"
        }
      ]
    },
    "keybindings": [
      {
        "command": "xxx.copyItemId",
        "when": "focusedView == xxxExplorer",
        "key": "ctrl+c",
        "mac": "cmd+c"
      }
    ]
  }
}
```

2. Select one of item in tree view `xxx-explorer`
3. Press `Crtl+C`

The command `xxx.copyItemId` executed but with empty parameters.

As same time clicking menu of `xxx-explorer` works well with correct parameter(tree item data).

**Comment by connor4312** [comment]
@TAUnionOtto this issue is for the Test Explorer view and is not related to custom views. Please open a new issue so your problem gets the right attention.

**Comment by connor4312** [comment]
We don't have any precedence got this thing. Executing a keybinding is the same as running it through the command palette, it's not contextual. I think if we want to solve this, we should do so in a generalizable way. Perhaps we allow the `when` clause on keybindings to include some new `activeTree` variable and, if present and that tree is currently focused, we pass its items to the command?

In core we'd want to make this opt-in so that all of a sudden all tree view internals aren't exposed to the outside world. Thoughts @jrieken @eamodio ?

**Comment by jrieken** [comment]
> Perhaps we allow the when clause on keybindings to include some new activeTree variable and, if present and that tree is currently focused, we pass its items to the command?

So, you mean only invoke that command when the activeTree is mine and only than pass the focused/selected element(s)?

**Comment by connor4312** [comment]
More like

- When a keybindings is fired
- If the keybinding's `when` clause includes an `activeTree`
- And a tree with ID `activeTree` is visible, has selected items, and 'opted-in' to this functionality
- Then pass the tree items to the command when it's invoked

**Comment by jrieken** [comment]
That's not how when-clauses are meat to work  - they aren't "doers" but just conditions that are checked

**Comment by alex-from-intuita** [comment]
+1 
we have a tree view for our extension to let folks go through a few code transformation recommendations and very quickly accept or reject those with the keyboard shortcut (e.g. type A for Accept and R for Reject).
 
is there any workaround/solution for us?
@jrieken , @connor4312 

![Screen Shot 2022-08-24 at 12 04 50 PM](https://user-images.githubusercontent.com/78109534/186502113-c72bb5bd-677e-47fc-b4f8-1696863ddc3b.png)

**Comment by eamodio** [comment]
I agree with @jrieken that the `when` clause isn't the right trigger to change the behavior of the callback, but just having an `activeView` (or more accurately `focusedView` `when` condition would help this scenario.

Because then you could register specific commands for each view you wanted to allow keybindings on -- and then in the command you'd just need to go and take action on the appropriate view.

Another addition related to this would be have still have the above, but also a way to get the "focused" view (if there is one), just like a text editor. That way you could then register a command with a set of view ids (via regex) and then map to the right one. It would also be great to get an event when the focused view changed.
